### PR TITLE
added ilr-phylogenetic and merged add-pseudocount

### PR DESCRIPTION
**Pull request by qiyunzhu**
Addressed issue #41 @mortonjt 

Brief summary of the Pull Request, including any issues it may fix using the GitHub closing syntax:

https://help.github.com/articles/closing-issues-using-keywords/

Also, include any co-authors or contributors using the GitHub coauthor tag:

https://help.github.com/articles/creating-a-commit-with-multiple-authors/

---

Include any questions for reviewers, screenshots, sample outputs, etc.

### CODE REVIEW

**Comment by mortonjt** [comment]
I approve!

**Review by mortonjt — APPROVED** [review]

File: q2_gneiss/composition/_composition.py [hunk 1]
```diff
@@ -56,3 +56,29 @@
...
+                 'partitions of features.  Each tip in the hierarchy'
+                 'corresponds to the feature identifiers in the table. '
+                 'This tree can contain tip ids that are not present in '
+                 'the table, but all feature ids in the table must be '
+                 'present in this tree.  This assumes that all of the '
+                 'internal nodes in the tree have labels.')
+    },
+    parameter_descriptions={},
+    output_descriptions={'balances': ('The resulting balances from the '
+                                      'ilr transform.')},
```
Comment (ebolyen) [inline q2_gneiss/composition/_composition.py]: Can we add a description for the tree? I assume this tree has the new internal node labels right?

File: q2_gneiss/composition/_method.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+                     pd.DataFrame):
+    return ilr_transform(table, tree)
+
+
+def ilr_phylogenetic(table: pd.DataFrame, tree: skbio.TreeNode) -> (
+                     pd.DataFrame, skbio.TreeNode):
+    t = tree.copy()
+    t.bifurcate()
+    t = rename_internal_nodes(t)
+    return ilr_transform(table, t), t
```
Comment (ebolyen) [inline q2_gneiss/composition/_method.py]: :+1:

File: q2_gneiss/composition/tests/test_method.py [hunk 3]
```diff
@@ -34,6 +35,28 @@ def test_ilr(self):
...
+    def test_ilr_phylogenetic(self):
+        np.random.seed(0)
+        table = pd.DataFrame([[1, 1, 2, 2],
+                              [1, 2, 2, 1],
+                              [2, 2, 1, 1]],
+                             index=[1, 2, 3],
+                             columns=['a', 'b', 'c', 'd'])
+        table = table.reindex(columns=np.random.permutation(table.columns))
+        tree = TreeNode.read([
+            '((c:0.025,d:0.025,e:0.025):0.2,(b:0.025,a:0.025):0.2);'])
```
Comment (ebolyen) [inline q2_gneiss/composition/tests/test_method.py:81]: :+1: for testing the bifurcation

File: q2_gneiss/composition/_composition.py [hunk 4]
```diff
@@ -56,3 +56,29 @@
...
+    description="Calculate balances given a hierarchy."
```
Comment (ebolyen) [inline q2_gneiss/composition/_composition.py]: "Calculate balances given a rooted phylogeny."

File: q2_gneiss/composition/_composition.py [hunk 5]
```diff
@@ -56,3 +56,29 @@
...
+
+
+plugin.methods.register_function(
+    function=ilr_phylogenetic,
+    inputs={'table': FeatureTable[Composition],
+            'tree': Phylogeny[Rooted]},
+    outputs=[('balances', FeatureTable[Balance]),
+             ('tree', Phylogeny[Rooted])],
+    parameters={},
+    name='Isometric Log-ratio Transform',
```
Comment (ebolyen) [inline q2_gneiss/composition/_composition.py]: Let's give these slightly different names

**Review by ebolyen — CHANGES_REQUESTED** [review]
Looks good, just a couple of minor things in the registration.

@mortonjt or @qiyunzhu, would either of you be able to create a corresponding docs PR for the gniess tutorial? Thanks!

**Comment by qiyunzhu** [comment]
Hi @ebolyen Thanks for your feedbacks! Regarding docs, I think Lisa is working on them in a separate PR. @mortonjt is that right?

**Comment by mortonjt** [comment]
Yup. That's correct

File: q2_gneiss/composition/_composition.py [hunk 6]
```diff
@@ -56,3 +56,29 @@
...
     description="Calculate balances given a hierarchy."
 )
+
+
+plugin.methods.register_function(
+    function=ilr_phylogenetic,
+    inputs={'table': FeatureTable[Composition],
+            'tree': Phylogeny[Rooted]},
+    outputs=[('balances', FeatureTable[Balance]),
+             ('tree', Phylogeny[Rooted])],
```
Comment (mortonjt) [inline q2_gneiss/composition/_composition.py]: Can we convert this tree to a hierarchy so that the regression summaries don't crash downstream?

**Comment by qiyunzhu** [comment]
travis build was stalled, any idea? @mortonjt

**Comment by qiyunzhu** [comment]
Done! @mortonjt @ebolyen

**Review by ebolyen — APPROVED** [review]
LGTM! Merge-pending Docs PR so that our CI system can still test other things.

File: q2_gneiss/composition/_composition.py [hunk 7]
```diff
@@ -5,54 +5,66 @@
...
-    function=add_pseudocount,
-    inputs={'table': FeatureTable[Frequency]},
-    parameters={'pseudocount': Int},
-    outputs=[('composition_table', FeatureTable[Composition])],
+    function=ilr_hierarchical,
+    inputs={'table': FeatureTable[Frequency],
+            'tree': Hierarchy},
+    outputs=[('balances', FeatureTable[Balance])],
+    parameters={'pseudocount': Float},
+    name='Isometric Log-ratio Transform',
```
Comment (lisa55asil) [inline q2_gneiss/composition/_composition.py]: Can you specify this to be specific for hierarchies?

For instance
`Isometric Log-ratio Transform applied to a hierarchical clustering`

File: q2_gneiss/composition/_composition.py [hunk 8]
```diff
@@ -5,54 +5,66 @@
...
-            'tree': Hierarchy},
-    outputs=[('balances', FeatureTable[Balance])],
-    parameters={},
+    function=ilr_phylogenetic,
+    inputs={'table': FeatureTable[Frequency],
+            'tree': Phylogeny[Rooted]},
+    outputs=[('balances', FeatureTable[Balance]),
+             ('hierarchy', Hierarchy)],
+    parameters={'pseudocount': Float},
     name='Isometric Log-ratio Transform',
```
Comment (lisa55asil) [inline q2_gneiss/composition/_composition.py]: Same thing here, can you change this to something like

`Isometric Log-ratio Transform applied to a phylogenetic tree`

**Comment by ebolyen** [comment]
Thanks @qiyunzhu! And everyone else who reviewed!

**Comment by qiyunzhu** [comment]
@ebolyen Thank you very much!

**Pull request merged**